### PR TITLE
Make local function parameters bind lazily

### LINKED ISSUE #16451: Compiler crash with recursive local function parameter default value
Typing the following code into a C# project causes VS to crash (stack overflow in the compiler).

```cs
class C
{
    public static void Main(int arg)
    {
        int Local(int x = Local()) => 2;
    }
}
```

### PR DESCRIPTION
**Pull request by agocke**
**Customer scenario**

Mentioning a local function in the default value of one of its parameters will cause the compiler to crash. The following program is the simplest repro for this scenario:

```csharp
class C
{
    public void M()
    {
        int Local(int x = Local()) => 0;
    }
}
```

**Bugs this fixes:** 

Fixes #16451

**Workarounds, if any**

Don't mention the local function in its parameters' default values.

**Risk**

This change preserves the existing code paths for binding of default parameters for almost all circumstances and simply attempts to delay the code path for local functions to occur later.

**Performance impact**

Low. Same code path, just done lazily.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

Insufficient testing across new language features and lack of adversarial testing. Unit tests have been added for this and scenarios like this.

**How was the bug found?**

Testing local function default parameters with other language features.

### CODE REVIEW

**Comment by agocke** [comment]
@dotnet/roslyn-compiler For review

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs [hunk 1]
```diff
@@ -194,23 +193,28 @@ internal override bool IsCallerMemberName
             }
         }
 
-        private ConstantValue DefaultSyntaxValue
+        private ConstantValue GetDefaultSyntaxValue(DiagnosticBag diagnosticsOpt = null)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs]: >private ConstantValue GetDefaultSyntaxValue(DiagnosticBag diagnosticsOpt = null) [](start = 8, length = 80)

Please open an issue to get rid of this pattern post RTM

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs [hunk 2]
```diff
@@ -67,12 +68,10 @@ private enum ParameterSyntaxKind : byte
...
 
             _lazyDefaultSyntaxValue = defaultSyntaxValue;
+            ParameterBinder = parameterBinderOpt;
         }
 
-        protected virtual Binder ParameterBinder
-        {
-            get { return null; }
-        }
+        protected virtual Binder ParameterBinder { get; }
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs]: >protected virtual Binder ParameterBinder { get; } [](start = 8, length = 49)

It feels like this property can be calculated from the containing LocalFunctionSymbol.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs]: >protected [](start = 8, length = 9)

private?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs (hunk 1 above)
Comment (agocke) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs]: Do you think this is encompassed by https://github.com/dotnet/roslyn/issues/16652?

That would remove the flakiness with adding to declaration diagnostics.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs]: Somewhat, I think we should explicitly track API changes what we would like to revert. We can keep the list in that issue.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs [hunk 3]
```diff
@@ -30,7 +30,19 @@ private enum ParameterSyntaxKind : byte
 
         private CustomAttributesBag<CSharpAttributeData> _lazyCustomAttributesBag;
         private ThreeState _lazyHasOptionalAttribute;
-        protected ConstantValue _lazyDefaultSyntaxValue;
+        private ConstantValue _lazyDefaultSyntaxValue;
+
+        private Binder _parameterBinder
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs]: >_parameterBinder [](start = 23, length = 16)

Does this follow naming convention for properties?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs [hunk 4]
```diff
@@ -30,7 +30,19 @@ private enum ParameterSyntaxKind : byte
...
+        {
+            get
+            {
+                var isLocalFunction = ContainingSymbol?.Kind == SymbolKind.Method &&
+                    ((MethodSymbol)ContainingSymbol).MethodKind == MethodKind.LocalFunction;
+                return isLocalFunction
+                    ? ((LocalFunctionSymbol)ContainingSymbol).ParameterBinder
+                    : null;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs]: While we typically prefer `Kind` checks, would it be simpler to use `as` instead here?
```
return (ContainingSymbol as LocalFunctionSymbol)?.ParameterBinder;
```

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs (hunk 3 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs]: Consider naming `LocalFunctionParameterBinder` so it's clear when the binder is expected.

File: src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs [hunk 5]
```diff
@@ -256,6 +256,70 @@ public void M()
...
+}");
+            comp.VerifyDiagnostics(
+                // (6,28): error CS1736: Default parameter value for 'p' must be a compile-time constant
+                //         int Local1(int p = Local2()) => 0;
+                Diagnostic(ErrorCode.ERR_DefaultValueMustBeConstant, "Local2()").WithArguments("p").WithLocation(6, 28),
+                // (7,28): error CS1736: Default parameter value for 'p' must be a compile-time constant
+                //         int Local2(int p = Local1()) => 0;
+                Diagnostic(ErrorCode.ERR_DefaultValueMustBeConstant, "Local1()").WithArguments("p").WithLocation(7, 28));
+            comp.DeclarationDiagnostics.Verify();
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs]: Is `nameof` interesting? `string Local(string s = nameof(Local)) => s;`

**Comment by cston** [comment]
LGTM

**Comment by agocke** [comment]
@MattGertz For escrow approval

**Comment by MattGertz** [comment]
Rejected for RTW per raised bar in Shiproom.  Please target 2.1 for this fix.

**Comment by AlekseyTs** [comment]
@agocke We probably shouldn't merge this into 2.1 as is without addressing the issue around GetDefaultSyntaxValue (https://github.com/dotnet/roslyn/pull/16736#pullrequestreview-18306789).

**Comment by agocke** [comment]
@AlekseyTs Agreed

**Pull request closed without merging**